### ISSUE TITLE
Building Hack on Fedora tracker

### ISSUE THREAD

**Issue by spstarr** [source-foundry/Hack#227]
Is this issue resolved? Fedora still does not have the fonts, I've taken over the packaging but we need a way to build these fonts to satisfy Fedora's policy:

**"Fonts SHOULD be built from source whenever upstream provides them in a source format"**

**Comment by chrissimpkins** [comment]
We do not have an automated build chain at the moment.  You can build them from the source with Font Forge.  How are you building other fonts?

**Comment by mavit** [comment]
DejaVu, for example, comes with a [Makefile](https://github.com/dejavu-fonts/dejavu-fonts/blob/master/Makefile) containing the recipe for building the fonts.  I see it calls a number of [scripts](https://github.com/dejavu-fonts/dejavu-fonts/tree/master/scripts) to do the work, including several that call FontForge (e.g., [`generate.pe`](https://github.com/dejavu-fonts/dejavu-fonts/blob/master/scripts/generate.pe)).  If Hack is based on DejaVu, maybe these can be largely reused?

**Comment by davelab6** [comment]
Its easy to automate fontforge builds with FF script (PE) or Python

**Comment by spstarr** [comment]
This is Fedora's full build for DejaVu fonts:
http://pkgs.fedoraproject.org/cgit/rpms/dejavu-fonts.git/tree/

**Comment by spstarr** [comment]
as @mavit said, if we can use the method DejaVu uses, then the Makefile in the DejaVu fonts can be adapted for Hack.

**Comment by mcepl** [comment]
> Its easy to automate fontforge builds with FF script (PE) or Python

Then I guess you would have no problems to provide PR with those scripts, right? What I mean is that we shouldn’t just require upstream to fix our problems, but we should help.

**Comment by spstarr** [comment]
We can build the fonts manually in the .spec I just didn't know since that's not documented. If the DejaVu fonts build is the same for Hack, then Fedora can adapt this in our packaging.

**Comment by chrissimpkins** [comment]
@spstarr still interest in this?  Any progress?

**Comment by spstarr** [comment]
Yes interested, just sidetracked in RL, are there any bugs I need to be aware of right now that impact building?

**Comment by spstarr** [comment]
I will be building today, finally.

**Comment by chrissimpkins** [comment]
>  just sidetracked in RL

Believe me, this was not a nudge.  I am the last person to do this given my own capacity to attend to the project recently.

> are there any bugs I need to be aware of right now that impact building

No Linux bugs that I am aware of that should cause issues for you

> I will be building today, finally.

👍  Congrats!

When you have a chance, can I ask you to provide me with how you would like to handle Fedora build associated questions when/if we receive them in this repo?

**Comment by chrissimpkins** [comment]
@spstarr we should also update the [README](https://github.com/chrissimpkins/Hack#package-managers) with this information when you are ready to go.  For Fedora we currently include the following package manager instructions:

`dnf-plugins-core :: heliocastro/hack-fonts :: hack-fonts`

**Comment by spstarr** [comment]
Yes, I would like to work with you to ensure we need no custom patches in Fedora, so any new versions you drop, i can build and push in (or see about any automated build/add to rawhide)

**Comment by chrissimpkins** [comment]
@spstarr you've scripted the build process?  We changed our release directory structure to facilitate releases on a different Linux distro.  Let me know if there is anything that I can do on our end to help you with the Fedora releases.

**Comment by chrissimpkins** [comment]
Adding this information that was discussed in [another IR](https://github.com/chrissimpkins/Hack/issues/35#issuecomment-305985407) and in [the Haack fork repo](https://github.com/ignatov/Haack/issues/4) re: hinting in the fonts based upon the title of this thread:

For ttf files (recommended for use across all platforms), I built with FontLab V as “pre-hinted” ttf build files, then used our ttfautohint based scripted hinting approach that is contained in the following script file:

https://github.com/chrissimpkins/Hack/blob/master/postbuild_processing/tt-hinting/autohint.sh

This uses hinting settings that are defined in the -TA- files in this directory:

https://github.com/chrissimpkins/Hack/tree/master/postbuild_processing/tt-hinting

These files apply optimized hinting to the main character set files that are used in code. We never ran through optimizations for extended character sets that were not commonly used in code because this was not our goal. If you replicate this approach, you will get hinting in your new ligature characters (ttfautohint hints all characters in the set) but you may need to tweak these if you find that they do not look as intended at certain glyph sizes.

For otf files, the build files are hinted with the built in otf hinting used in FontLab V. You will get good automated hinting of otf files in Glyphs as well (probably better c/w FontLab).

On OS X, the hinting is automated by the operating system and hints are ignored. TTF and OTF files should both look similar and good.

For Windows and Linux, the hints really, really matter (see a large number of the closed issue reports) and I suggest that you investigate how your new glyphs appear at a range of sizes commonly used in code editors if you are targeting these platforms.

**Comment by spstarr** [comment]
I started, a little, today more... I'll put up a .spec file on my fedora page hopefully today.

**Comment by chrissimpkins** [comment]
@spstarr let me know if there is anything you need from us.  Be happy to explain the hinting process if it would be helpful to you.  Thanks Sean.

**Comment by spstarr** [comment]
@chrissimpkins I'm following all your changes right now I see 3.0 is coming but not yet. I'll have a spec file this week but not against 3.0.

**Comment by chrissimpkins** [comment]
@spstarr note the change in source structure if you are pulling source for your build.  Once our build chain is in place this may be something you want to convert to.  Part of the goal of this effort so that we offload this from you!

**Comment by chrissimpkins** [comment]
@spstarr OK if I hijack this thread to use for the general build toolchain conversation?  Happy to create a new one if you want this to be Fedora specific.

**Comment by spstarr** [comment]
No problem this is fine

**Comment by chrissimpkins** [comment]
@spstarr great thanks.  more updates re build chain coming this weekend

**Comment by chrissimpkins** [comment]
build dependencies will include Python 2.7 or Python 3.5+

**Comment by chrissimpkins** [comment]
desktop ttf file build dependencies will include:

- fontmake (Python, available on PyPI `pip install fontmake`)
- ttfautohint v1.6+ (C, https://www.freetype.org/ttfautohint/)

**Comment by chrissimpkins** [comment]
New functionality was introduced in ttfautohint 1.6 that we will be using in our manual hints.  v1.6+ will be mandatory for ttfautohint.  fontmake does not necessarily need to be pinned to the current version at this stage.  I will keep an eye on the fontmake project changes and update if they introduce something that we anticipate to break our builds in some fashion.

**Comment by chrissimpkins** [comment]
@spstarr @paride 

Sean and Paride, 

I added a new `build.sh` script in the top level of the repository that currently builds unhinted ttf files from the UFO source.  It includes a `--install-dependencies` flag to install the build dependencies in an optional fashion.  Would the two of you be willing to have a look at this script and see if this dependency management is appropriate for your purposes or if you intend to perform these installs yourself prior to the build for your users?

Feel free to PR in any recommended changes that will make this better serve your purposes for a source based build/install on your respective Linux distros, ideally maintaining an approach that also can be used by anyone who would like to clone the repository and build the files with the script in order to manually install them independent of a package manager.

The script is currently located in the `dev-build-scripts` branch here:

https://github.com/source-foundry/Hack/blob/dev-build-scripts/build.sh

I will be updating the script with hinting and a couple of new scripts to address post build fixes that are required over the course of this week.  The build files are released on the path `build/ttf/Hack-*.ttf` and will remain that way with the hinting and post build fixes that are yet to be implemented.

**Comment by chrissimpkins** [comment]
@burodepeper 

David,

If you happen to have time this week, would you be willing to give the scripted dependency install a try on your OS X box.  This will set up the dependencies for you so that you can build the desktop fonts for testing after any changes that you make.

Let me know if you come across any issues.

**Comment by chrissimpkins** [comment]
ttfautohint scripting added to the build.sh script.  This currently applies automated instruction sets to the fonts with ttfautohint so the files that are on the path `build/ttf/Hack-*.ttf` are now considered "hinted"; however, there are outstanding manual hinting adjustments that are required before release.   

See issues #258 #259 #260 #261 for the hinting issues that will be addressed for the v3.0 release.

This issue is on my plate.  Will update in those threads when it is complete.

**Comment by burodepeper** [comment]
@chrissimpkins Ran into the following issues:

```
configure: error: Cannot find qmake.  Try --with-qt=PATH.
sudo: ttfautohint-1.6/make: command not found
./build.sh: line 38: [: missing `]'
./build.sh: line 44: installflag: command not found
/usr/local/bin/fontmake
Unable to install ttfautohint from source.  Please attempt manual install and repeat build without the --install-dependencies flag.
./build.sh: line 63: =: command not found
./build.sh: line 67: [: -eq: unary operator expected
./build.sh: line 138: ttfautohint: command not found
Unable to execute ttfautohint on the Hack-Regular variant set.  Build canceled.
```

No idea what `qmake` is. Haven't tried manually installing `ttfautohint` yet. The other errors are syntax errors?

**Comment by chrissimpkins** [comment]
@burodepeper Thanks David.  This is helpful.

> configure: error: Cannot find qmake.  Try --with-qt=PATH.

It looks like ttfautohint's default install requires Qt to support its GUI version.  

@lemzwerg (re: error report in post above https://github.com/source-foundry/Hack/issues/227#issuecomment-322166712) is it possible to build *only* the command line version of ttfautohint from source (i.e. without the GUI and the Qt dependency)?  We are creating a scripted build process for our fonts and need to implement ttfautohint as part of the build chain.  There are a few Linux distros that would like to build the fonts from source in order to remain within their software release guidelines/regulations.  Given that we are putting this into motion, I would like for anyone on Linux or OS X platforms to be able to build the fonts from source but we don't need ttfautohint GUI, just the command line executable for the shell script.  Would be great to eliminate the Qt dependency for this if there is a way to do so.

**Comment by chrissimpkins** [comment]
> ./build.sh: line 38: [: missing `]'

This is a syntax error and my fault.  Need a space before the closing bracket.

**Comment by chrissimpkins** [comment]
> ./build.sh: line 44: installflag: command not found

Syntax error.  Need to change:

```
installflag = 0
````

to 

```
installflag=0
```

**Comment by chrissimpkins** [comment]
> ./build.sh: line 63: =: command not found

I need to remove the spaces in the variable assignments for each of the install confirm blocks to address this.

**Comment by chrissimpkins** [comment]
> ./build.sh: line 67: [: -eq: unary operator expected

I think that this is simply because the variable was never assigned an integer value because of above assignment issue.  I believe that the above fixes will take care of it.

**Comment by chrissimpkins** [comment]
> Besides it all, it did generate a set of .ttf files in /master_ttf which appear to be valid at first glance.

OK, you are installed through the fontmake part of the dependencies then.  The script didn't complete because `master_ttf` is a temporary directory.  Those are unhinted builds from the UFO source.  Following that step they will be hinted with ttfautohint, a few other (yet to be built) modifications of the font binaries, and then moved for release to the `build/ttf` directory.

**Comment by lemzwerg** [comment]
@chrissimpkins, a call to ttfautohint's `configure --help` would tell you the following.
```
--with-qt[=ARG]         Qt support. ARG can be `yes' (the default), `no', or
                        a path to Qt binaries; if `yes' or empty, use PATH
                        and some default directories to find Qt binaries
```
In Homebrew, if you omit the `--with-qt5` option, only the command line version gets built.

**Comment by chrissimpkins** [comment]
@lemzwerg perfect!  Thanks much Werner.  Sorry, I didn't check the in app help string, checked the website documentation.  Possible for me to add this information via PR to your OS X build page?

**Comment by lemzwerg** [comment]
@burodepeper, the standard incantation for building almost all Unix packages from source is
```
./configure <options>
make
make install
```
A call to `./configure --help` shows you the possible configuration options.

**Comment by chrissimpkins** [comment]
Will need to modify line https://github.com/source-foundry/Hack/blob/dev-build-scripts/build.sh#L32 to this:

```
ttfautohint-1.6/configure --with-qt=no
```

based upon Werner's post above.

**Comment by burodepeper** [comment]
@lemzwerg Thanks! I'm a GUI-loving, pre-build binary kind of nerd... ;)

**Comment by lemzwerg** [comment]
@chrissimpkins, yes, please do so – your English is certainly better than mine :-)
You can get the FreeType web stuff via [git](https://anongit.freedesktop.org/git/freetype/freetype-web.git/); if you have a patch, please send it to me by e-mail.

**Comment by spstarr** [comment]
@chrissimpkins I am going to be reviewing https://bugzilla.redhat.com/show_bug.cgi?id=1258542 which is dependency for building Hack fonts.

**Comment by spstarr** [comment]
ttfautohint is currently version 1.6 in Fedora

**Comment by chrissimpkins** [comment]
> You can get the FreeType web stuff via git; if you have a patch, please send it to me by e-mail.

@lemzwerg will do!  Will be in touch.  Thanks again for your help Werner.  I really appreciate it.

**Comment by chrissimpkins** [comment]
@spstarr @paride Can you verify that the other dependencies in Werner's documentation are available through your package manager or are part of the default distro install?

https://freetype.org/ttfautohint/osx.html

I am guessing that FreeType and pkg-config are installed by default on Linux distros?  Harfbuzz likely requires separate install?  We won't need the Qt dependency (see https://github.com/source-foundry/Hack/issues/227#issuecomment-322186294)

**Comment by chrissimpkins** [comment]
build script updated to install ttfautohint with Homebrew on OS X platform.  This automates the install of all ttfautohint build dependencies on OS X.  Await feedback from Linux groups to see if they have suggestions about how to approach this on various Linux distros.  Package managers are likely going to be the answer if they have necessary versions of the Hack build dependencies available.  

@burodepeper should be able to give this another shot on OS X to setup your development environment.  Let me know if you see any other issues.  All original reported issues are fixed with the latest commit.

**Comment by spstarr** [comment]
@chrissimpkins freetype, pkg-config (and the new replacement of pkg-config) are in Fedora, some dependencies are implicit and get pulled in, when I build things i'll know whats missing. Fedora has Harfbuzz v1.4.4, Freetype 2.7.1. I would like to use the package management dependency install, all the build/runtime dependencies should be in Fedora so it makes for an easy build experience. 

Fedora also can keep track of new version releases and notify us and or even build a new version when you release :)

**Comment by chrissimpkins** [comment]
@spstarr thank you!  All sounds good.  Assuming that this will be the same with other Linux package managers (as it should be) and that this will be the simplest approach on both of our parts, will plan for the `--install-dependencies` flag to serve the purpose of installs from source for end users who want to build directly from source (testers, potential contributors) or who want to modify the font source in some way and build in their modifications.  We are attempting to support these needs in addition to Linux distro specific redistribution with the scripted builds.

**Comment by spstarr** [comment]
That sounds like a good solution, developers that want to build things local and pull in dependencies can use `--install-dependencies`. It also is useful for packagers in seeing what new dependencies are required. I typically try to build a program and see what fails when it's newer than what is packaged then add in any new dependencies into package if they exist in the distribution.

**Comment by chrissimpkins** [comment]
:+1:  We will add build documentation as well.  Let your fellow Fedorians know that their suggestions were heard ;)

**Comment by mavit** [comment]
> This is a syntax error. Need a space before the closing bracket.

By the way, this is the kind of problem can be picked up with, say, [ShellCheck](https://www.shellcheck.net/).  It's currently reporting a number of errors like the following:

```
In build.sh line 72:
                            $installflag=1
                             ^-- SC1066: Don't use $ on the left side of assignments.
```

ShellCheck can be [integrated into Travis](https://github.com/koalaman/shellcheck/wiki/TravisCI), if you're interested.

Hope this helps.

**Comment by chrissimpkins** [comment]
You mean I will need to look at my poor shell scripting skills with every commit??!! ;)

That is fantastic and would be very helpful across all of our shell scripts Peter.  I wasn't aware of these shell scripting linter projects.  I am dusting off the ole' shell scripting text as I dig back into these files.  

Something that you would like to contribute to the project (implementation of Travis testing on the build.sh script)?  We would be happy to have the contribution if you have time/interest.  This script won't be in master until we release v3.0 but it will be a very important component of that release and I really like the idea of CI testing as necessary modifications to the file come up.

**Comment by chrissimpkins** [comment]
@mavit created a new IR for shell script CI testing https://github.com/source-foundry/Hack/issues/269

**Comment by chrissimpkins** [comment]
All issues identified in build.sh by ShellCheck were fixed in commit https://github.com/source-foundry/Hack/commit/a3dbb96beb0273f00695715188ea521d369f195f

**Comment by chrissimpkins** [comment]
The scripting for builds to desktop ttf files is complete and functional as of https://github.com/source-foundry/Hack/commit/feeb927fa63f6c08936123026f083f9f529629a9 

I still need to update the manual hinting for these builds (and they should not currently be considered release ready installable files), but this an issue external to the build.sh shell script and should not require modification of the file.

OK for package maintainers to use build.sh to evaluate for build dependencies at this stage.  There have been no additions beyond what was previously discussed in this thread.  There will be additional build dependencies for web fonts but I intend to use a separate script for that build process so this will not affect those who are building desktop ttf releases only.

Build documentation will be developed https://github.com/source-foundry/Hack/issues/265

**Comment by chrissimpkins** [comment]
Apologies, Harfbuzz > 0.9.19 is a dependency for the ttfautohint source build.  I am attempting to add this as part of the --install-dependencies installation in the build.sh script for Linux users (OS X can use Homebrew installs which is how this is implemented in script) but failing for now on travis...

**Comment by chrissimpkins** [comment]
It looks like Harfbuzz at appropriate version level is available in most Linux package managers.  Here are their build docs:

https://github.com/behdad/harfbuzz/blob/master/BUILD.md

**Comment by chrissimpkins** [comment]
@behdad I'm sure that this is something I am missing on my end but do you happen to have any thoughts on our Harfbuzz install process here?  I am trying to script an approach to install dependencies for ttfautohint as part of our font compilation process.  Per Werner's documentation, ttfautohint requires Harfbuzz v0.9.19+.  All Harfbuzz dependencies are added to the Travis .yml file via the apt install instructions prior to build according to your build documentation and the configure step executes without issues.  It appears to hang at the make step.

Here is the start of the Harfbuzz install in our build shell script: https://github.com/source-foundry/Hack/blob/dev-build-scripts-shellCItest-patch/build.sh#L51 

Here is what I am getting in Travis: 

```
./build.sh: line 54: harfbuzz-1.4.8/make: No such file or directory
```

and Travis stdout/err link: https://travis-ci.org/source-foundry/Hack/jobs/265673968

Sorry to waste your time, but I am stuck.

**Comment by behdad** [comment]
`harfbuzz-1.4.8/make` is wrong.  `make` is a separate command.  You just run it.

**Comment by lemzwerg** [comment]
Here's a script that builds ttfautohint (but not ttfautohintGUI) without any other dependencies – no need for Macports or Homebrew.  It also downloads the necessary packages (ttfautohint, freetype, harfbuzz).  Maybe it helps.

[ttfautohint-build.zip](https://github.com/source-foundry/Hack/files/1232260/ttfautohint-build.zip)

**Comment by chrissimpkins** [comment]
@behdad 

> harfbuzz-1.4.8/make is wrong. make is a separate command. You just run it.

Not necessary to be in the root of the project directory where it needs to be executed?  This is my misunderstanding of make if not.  The execution is tied to the previous `./configure` command?  Sorry for the ignorance on this end.  Shell scripting and these build processes are not a forté.

**Comment by chrissimpkins** [comment]
@lemzwerg 

> Here's a script that builds ttfautohint (but not ttfautohintGUI) without any other dependencies – no need for Macports or Homebrew. It also downloads the necessary packages (ttfautohint, freetype, harfbuzz). Maybe it helps.

👍 

Thanks so much Werner!  I will try to plug this in to the build chain this evening without breaking your script :)

**Comment by chrissimpkins** [comment]
@lemzwerg Possible to add a license to this?

**Comment by chrissimpkins** [comment]
I was missing the steps of directory changes into the harfbuzz source root...

From Werner's script:

```
cd "$HARFBUZZ"

./configure \
  --disable-dependency-tracking \
  --disable-gtk-doc-html \
  --with-glib=no \
  --with-cairo=no \
  --with-fontconfig=no \
  --with-icu=no \
  --prefix="$INST" \
  --enable-static \
  --disable-shared \
  CFLAGS="$TA_CPPFLAGS $TA_CFLAGS" \
  CXXFLAGS="$TA_CPPFLAGS $TA_CXXFLAGS" \
  LDFLAGS="$TA_LDFLAGS" \
  PKG_CONFIG=true \
  FREETYPE_CFLAGS="-I$INST/include/freetype2" \
  FREETYPE_LIBS="-L$INST/lib -lfreetype"
make
make install
cd ..
```

**Comment by lemzwerg** [comment]
LIcense of the script: Public domain

**Comment by chrissimpkins** [comment]
@lemzwerg ty!

**Comment by paride** [comment]
@chrissimpkins thanks for your work on the script. On `--install-dependencies`: the Debian policy requires that software in the main distribution does not depend or build-time depend of software that is not in the main distribution. This means that `fontmake` and its dependencies will have to be Debian packaged (currently it is not), which is fine, but this will require a non trivial effort. First let's see if we can settle the licensing issue.

The `fonttools` and `ttfautohint` 1.6 are already Debian packaged.

**Comment by chrissimpkins** [comment]
@paride are typefaces that build from source on Debian using `fontforge`?  Is this in main?

**Comment by paride** [comment]
@chrissimpkins I'm not sure about how other typefaces are built, but `fontforge` is definitely in main:

https://packages.debian.org/search?keywords=fontforge

(That's not the latest version yet, but I'm confident it will be packaged soon.)

**Comment by chrissimpkins** [comment]
@paride Thanks Paride.  Can you point me to the specific guideline that states this and/or any other Debian education materials that are available to package developers / maintainers and provide guidance?  I would like to educate myself on the matter a bit so we are on the same page with this discussion.

On our end, the need to support two different compilers is not ideal but I understand the issue from your standpoint.  There is going to be a bit of a rabbit hole with dependencies for fontmake and I don't know how deep that leads for inclusion in Debian main.  Since you include the fontTools library you will already have some of them included.  Probably an issue that you would want to discuss with the [fontmake team](https://github.com/googlei18n/fontmake).  My guess is that they would be very receptive to inclusion in Debian main in order to access the Debian/Ubutntu distros as a general purpose font compiler for other projects that may face this same issue with the DFSG.

**Comment by anthrotype** [comment]
Thanks for pinging me @chrissimpkins.
Of course we'd be happy if fontmake and its dependencies were packaged by Debian, although I am not familiar with the process involved in getting packages into Debian or other Linux distros.
Note that fontmake & co. are simple python packages, they only require a decent version of setuptools to build. All their inter-dependencies are specified in each package's `install_requires` field of `setup.py`, so that pip can resolve them when installing from the Python Package Index.
Please let me know if I need to do anything to facilitate this process.

**Comment by chrissimpkins** [comment]
@anthrotype thanks for your note Cosimo.  @paride manages the Hack package for Debian / Ubuntu and is the contact for this issue.   Will let him weigh in on whether this might be something that he builds or if there are others who would join the effort.

**Comment by paride** [comment]
@chrissimpkins The Debian New Maintainers' Guide is here:

https://www.debian.org/doc/manuals/maint-guide/

but I warn you: it's very long and detailed. You can also browse the Debian Wiki, for example:

https://wiki.debian.org/DebianMentorsFaq

Here is the Debian Policy:

https://www.debian.org/doc/debian-policy/

where, for example, you'll find stated that Debian packages in the "main" distribution can't (build-)depend on packages that are not in the "main" distribution themselves. This is stated here:

https://www.debian.org/doc/debian-policy/ch-archive.html#s-main

The Debian Social Contract is probably the most important document for Debian:

https://www.debian.org/social_contract

and it's the one that contains the Debian Free Software Guidelines (dfsg):

https://www.debian.org/social_contract#guidelines

**Comment by chrissimpkins** [comment]
@paride :+1:  tyvm!   let me educate myself over the weekend and will be back in touch

**Comment by paride** [comment]
@chrissimpkins on `fontmake`/`fontforge`: there is already a "request for package" (RFP) issue for it:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=865283

some dependencies are missing indeed (at least otf2ft, glyphsLib, booleanOperations), but it doesn't look so terrible, and we don't need to package Hack 3.0 from day zero. What I'm saying it: use the best tool for the job, do not duplicate your efforts. As long as you're using only free software we'll sort it out.

**Comment by chrissimpkins** [comment]
@paride : 

> What I'm saying is: use the best tool for the job, do not duplicate your efforts. As long as you're using only free software we'll sort it out.

:+1:  Free, open source software only for all build steps from UFO to all final fonts (including web fonts) is the plan as of the v3.0 release.  Thanks for the ongoing feedback Paride.  It is extremely helpful.

**Comment by chrissimpkins** [comment]
@anthrotype 

> Sorry if this may sound naïve to you, but I don't quite understand why it would be ok to have a shell script that downloads the source tarballs for, say, harfbuzz or freetype in order to compile ttfautohint

We kind of got sidetracked in this issue report thread with the Debian package conversation.  My understanding is that all of the compiled applications that you mentioned are in Debian main.  Of the Python applications/library dependencies, fontTools is the only one in main at the moment.

If you are referring to the `build.sh` shell script in development here, the `--install-dependencies` flag is intended for local installs of build dependencies and to my knowledge there are no plans by any of the Linux distro package maintainers to use our script's `--install-dependencies` flag (or our build script as released) to manage their packages.  They intend to use their distro specific package managers.  Our `--install-dependencies` flag in the build script is intended for end users who would like to build from source, including those who might want to set up a development environment in order to contribute back upstream or create their own fork of the project.  The Linux specific conversations have taken place as part of the dependency discussion during our work on the build workflow.

> Why is it so important for Debian to have fontmake and all its dependencies in the "main" distribution?

This is something that I do not understand either.  My limited understanding is that they must review the compile tools against the Debian Free Software Guidelines in order for them to be considered acceptable for the build processes used in the distribution of software.  Paride linked a set of Debian specific guidelines/guidance links in https://github.com/source-foundry/Hack/issues/227#issuecomment-323434614 if you'd like to learn more.  Of the Linux distros that we've worked with in this project, Debian (including Ubuntu) seems to have the most stringent set of guidelines for free software re-distribution, including these build issues which were  to me as well.

> I struggle to see why I should invest in that 

Hopefully, there will not be any investment whatsoever on your part.  All of our Linux package maintainers released their respective packages completely independent of the involvement of our project team.  We helped with licensing questions, some organization of archives, and information about our release structure but beyond that I have no idea what they did to release our project on their respective distros.  I am guessing an interested Debian user will take the reins on your project and run with it.  It may just need a bit of a nudge from their users who have an interest in typefaces that cannot be released on Debian/Ubuntu because of the issue identified here.

**Comment by chrissimpkins** [comment]
New scripted build with inclusion of Werner's ttfautohint + ttfautohint dependency install script is now passing Travis tests.  https://github.com/source-foundry/Hack/commit/19a2b9390e49d7db93a90e04e05fe51b3e79b187 

Werner's script performs a local ttfautohint install on the path `$HOME/ttfautohint-build/local/bin/ttfautohint`.  I've included a test to determine whether ttfautohint is installed on PATH if it is not found on the path `$HOME/ttfautohint-build/local/bin/ttfautohint` so that users have the option to use a version that is installed on their PATH (e.g. from previous install) if they are skipping the optional scripted dependency installs in the build.sh script.

**Comment by mavit** [comment]
> Why is it so important for Debian to have fontmake and all its dependencies in the "main" distribution?

I'm not a member of Debian and can't speak for them, but I think the main reason is as follows: if you're building an operating system that you hope to be able to support for a decade or more into the future, it's important to be confident that you have everything you need to be able to modify and rebuild all of it's components.  

A build process can't be make to work reliably if it depends on pulling stuff in from random websites.  Websites go down temporarily or permanently, projects move their hosting, companies cease trading or [get bought out and become disreputable](https://en.wikipedia.org/wiki/SourceForge#Project_hijackings_and_bundled_malware).  These are rare events per package, but common events when multiplied by the number of packages in a Linux distribution.  So, instead, they need to take a copy of all of the dependencies, and integrate them into their distribution.

**Comment by chrissimpkins** [comment]
The first version of build documentation is available.  Discussion thread: https://github.com/source-foundry/Hack/issues/265

**Comment by chrissimpkins** [comment]
web font build workflow that takes us from ttf to woff is now available.  Work in the `dev-build-scripts` branch.

- `build.sh` = UFO to ttf files (desktop)
- `build-web.sh` = ttf to woff (web fonts)

Still trying to decide whether to implement another step to release woff2.  Will check how much additional compression we get.

For those who intend to build web fonts, there is one additional compiled dependency at the moment:

- [sfnt2woff-zopfli](https://github.com/bramstein/sfnt2woff-zopfli)

As with the desktop build script, automated steps to pull and build the above project are available for users with the `--install-dependencies` flag.  As with ttfautohint (and all of its build dependencies) we are using a local install on the path $HOME/project for users who install build dependencies with this approach.  No executables are installed on PATH.

Web font discussion / build thread with additional details: https://github.com/source-foundry/Hack/issues/286

**Comment by chrissimpkins** [comment]
@lemzwerg 

Your ttfautohint-build.sh script works without issues locally on OS X for me (OS X 10.12.5, XCode 8.3.3); however, I have previously installed freetype and harfbuzz with Homebrew.

When I try to implement [your script](https://github.com/source-foundry/Hack/blob/dev-build-scripts-osx-CI-test/tools/scripts/install/ttfautohint-build.sh) on a [Travis CI build](https://travis-ci.org/source-foundry/Hack/jobs/268703666#L4784) with OS X (defaults to OS X 10.11, XCode 10.11), there is an issue with the harfbuzz build dependency install:

```
Package freetype2 was not found in the pkg-config search path
```

Here is the line in the test std err output https://travis-ci.org/source-foundry/Hack/jobs/268703666#L4784

This leads to a ttfautohint build fail and subsequently build failure across the remainder of the build steps in the workflow. 

Wondering whether I should fall back to a homebrew install for ttfautohint on OS X?  While homebrew is very widely used, it means an additional dependency.

**Comment by lemzwerg** [comment]
@chrissimpkins, you misread the log file AFAICS – harfbuzz is built just fine, however, ttfautohint makes problems.

I believe the reason for your problem is a bug in FreeType.  Please replace the empty `FREETYPE_PATCHES` line in `ttfautohint-build.sh` with the following.

```
FREETYPE_PATCHES="\
  http://git.savannah.gnu.org/cgit/freetype/freetype2.git/patch/?id=c9a9cf59 \
  http://git.savannah.gnu.org/cgit/freetype/freetype2.git/patch/?id=c8829e4b \
"
```

In case this doesn't fix the issue, I need the `config.log` files from the FreeType and ttfautohint builds (the former is located in `.../freetype2/builds/unix/`).

**Comment by chrissimpkins** [comment]
@lemzwerg 

You are correct, I misread the headers and did not see the ttfautohint start in the stdout stream.

The patches do not appear to fix the issue.  

https://travis-ci.org/source-foundry/Hack/jobs/268718405#L4081

Start of ttfautohint install standard streams from the ttfautohint-build.sh script: https://travis-ci.org/source-foundry/Hack/builds/268723414#L4367

Start of `cat $HOME/ttfautohint-build/freetype-2.8/builds/unix/config.log`: https://travis-ci.org/source-foundry/Hack/builds/268723414#L4916

Start of `$HOME/ttfautohint-build/ttfautohint-1.6/config.log` :  https://travis-ci.org/source-foundry/Hack/builds/268723414#L5715

Those links take a few seconds to get you to the proper lines, but they will get you there.   Sorry, not straightforward to copy the web output to copy/paste text.

**Comment by lemzwerg** [comment]
@chrissimpkins, one patch hunk gets rejected; line https://travis-ci.org/source-foundry/Hack/builds/268723414#L3648 shouldn't be present (the first fail for the `ChangeLog` hunk is expected).  Please insert `echo` commands to check why – maybe you aren't using GNU patch (I have version 2.7.5), and the used command line options are ignored or misinterpreted – it's actually an error on my side to rely on the GNU version in my script...

Please replace the `patch` call and its long option names with the POSIX compatible command
```
patch -p1 -N -r - < $i
```
Another reason of failure might be that you don't start with a clean `freetype-2.8.tar.gz` archive...

However, the above is just a side issue that should be investigated.  The source of your problems is something else, I now believe – attached is a new version of the script that tries harder to completely ignore `pkg-config`.

[ttfautohint-build.sh.zip](https://github.com/source-foundry/Hack/files/1254268/ttfautohint-build.sh.zip)

Please adapt it to your needs.

**Comment by chrissimpkins** [comment]
@lemzwerg Passing builds on both Linux and OS X with your changes.

The only modification that I made was to convert the build directory for FreeType, Harbuzz, and ttfautohint back to `ttfautohint-build` from `ttfautohint-build-linux` in order to make it compatible with the paths that we are using for these local installs elsewhere in the build workflow.

shellcheck linting on the script fails on some relatively minor issues that I will modify so that we pass these linting tests unless you feel that any should be ignored instead.

shellcheck report here:  https://travis-ci.org/source-foundry/Hack/jobs/268949490#L572

Thanks again for all of your help Werner!

**Comment by chrissimpkins** [comment]
We were previously using v1.6 of ttfautohint. This script bumps to v1.7 which was just released.  No changes that will affect Hack builds.

Changelog: https://sourceforge.net/projects/freetype/files/ttfautohint/1.7/

**Comment by chrissimpkins** [comment]
Werner's (very slightly modified) new script has been merged to `dev-build-scripts` branch https://github.com/source-foundry/Hack/commit/d488635791a07a821190de3e60d926deceb99cc5

**Comment by chrissimpkins** [comment]
@lemzwerg 

Werner, since you went to the work of pulling this script together, I added it to a separate repository so that others can use it to build ttfautohint, including all of its build dependencies, from scratch on both Linux and OS X.

https://github.com/source-foundry/ttfautohint-build

Builds from scratch on both platforms should now be as simple as:

```
$ git clone https://github.com/source-foundry/ttfautohint-build.git
$ cd ttfautohint-build
$ make
```

or by release:

```
$ curl -L -O https://github.com/source-foundry/ttfautohint-build/archive/v1.7.tar.gz
$ tar -xzvf ttfautohint-build-1.7.tar.gz
$ cd ttfautohint-build-1.7
$ make
```

and the installs for ttfautohint and all dependencies remain on the path `$HOME/ttfautohint-build` as you initially defined it.

I am currently running Travis CI testing on the script (tests execution of entire FreeType, Harfbuzz, and ttfautohint build to end of script, for presence of the ttfautohint executable on the defined install path at the completion of the build, and shellcheck + checkbashisms linting on the shell script) in that repo across Linux (Travis defaults to Ubuntu Trusty at the moment) as well as OS X 10.10 (XCode 6.4), 10.11 (XCode 7.3), 10.12 (XCode 8.3).

The script passes builds on all tested platforms/versions, except OS X 10.10.  This appears to be a gnulib issue with the `max_align_t` definition in the `stddef.h` header file on that release of OS X.

This is apparently something that is recognized in gnulib:

https://www.gnu.org/software/gnulib/manual/gnulib.html#stddef_002eh

For our purposes here in the Hack build, we will let users know that OS X 10.11+ is a build requirement on the OS X platform.  Just pointing this out in case you are interested in targeting users on OS X with releases earlier than 10.11 (10.10 was released 10/2014, 10.11 was released 09/2015).  

The gnulib errors during the ttfautohint portion of the build begin at this line in the OS 10.10 tests:

https://travis-ci.org/source-foundry/ttfautohint-build/jobs/269031319#L3401

I'd be happy to add you to the repository with write access or feel free to fork it if this is of benefit to you/your users.  I can delete the repository if you establish a fork.  Maintenance is clearly better in your hands than in my own.

**Comment by lemzwerg** [comment]
Thanks for repository and invitation!  I've joined source-foundry :-)
Note that I have a bunch of different versions of this script to support cross compilation, etc.  No idea whether I'll find time to clean everything up and commit it...

Regarding the `max_align_t` issue: I've sent a report to bug-gnulib@gnu.org so that it gets fixed soon, hopefully.

**Comment by lemzwerg** [comment]
@chrissimpkins, I need the `config.log` file of the failing build – the main question is why `configure` doesn't properly recognize clang's `max_align_t` implementation...

**Comment by chrissimpkins** [comment]
> Note that I have a bunch of different versions of this script to support cross compilation, etc. No idea whether I'll find time to clean everything up and commit it...

No worries.  It's out there and available as is.  There was definitely some effort to get to this stage and it works on the current OS X release.

> Regarding the max_align_t issue: I've sent a report to bug-gnulib@gnu.org so that it gets fixed soon, hopefully.

:+1: 

> I need the config.log file of the failing build – the main question is why configure doesn't properly recognize clang's max_align_t implementation...

Started a new issue report over there for this.  https://github.com/source-foundry/ttfautohint-build/issues/1  .  We should be able to add this as an `after_failure` script to the .travis.yml file.  

You should have admin access on that repo as well.

**Comment by spstarr** [comment]
@chrissimpkins 

So, Fedora is working on getting other dependencies ready for Hack however:

Core ticket for hack-fonts:  https://bugzilla.redhat.com/show_bug.cgi?id=1258542

Someone is packaging python-fontmake but this needs to pull in other packages, here is a list of all the dependencies this package needs:

Dependencies:  https://bugzilla.redhat.com/show_bug.cgi?id=1441023 (python-fontmake)
Sub-dependencies: https://bugzilla.redhat.com/show_bug.cgi?id=1440992 (python-booleanoperations)
-> https://bugzilla.redhat.com/show_bug.cgi?id=1440968 (python-setuptools_scm_git_archive)
-> https://bugzilla.redhat.com/show_bug.cgi?id=1440971 (python-pyclipper)

These are working their way though the process.

fonttools = fonttools-3.15.1-1.fc28 (rawhide), fonttools-3.14.0-1.fc27 (next release)
ttfautohint = ttfautohint-1.6-3.fc27 (next release), no rawhide build yet

I think this should be all the dependencies.

**Comment by chrissimpkins** [comment]
@spstarr 

Thanks for the update!  @anthrotype will be interested in this as you make progress on Fedora.  The Debian team is working through the same process.

 It will be a bit of effort, but well worth it in my opinion.  I suspect that right now, typeface developers (and their users) are limited to fontforge as the only available compilation tool on most Linux distros.  There is a great deal of active work out there on free open source typeface development tooling and it would be great to have most/all of these tools reviewed by the various Linux distros for inclusion according to their packaging guidelines.  I'm glad to hear that our project has provided an impetus and greatly appreciate all of your help guiding it through this process.  I am only beginning to understand the amount of work involved :)

Both @anthrotype and @davelab6 are good contacts for the breadth of OSS tools out there if anyone at Fedora (or for @paride, Debian/Ubuntu) is interested in a deep dive for typeface compilation / post build modification / font editors support on your distros.

**Comment by chrissimpkins** [comment]
> ttfautohint = ttfautohint-1.6-3.fc27 (next release), no rawhide build yet

This is acceptable.  Anything 1.6+ works for our builds.  We haven't pinned the other dependencies at any specific version.  Please let me know if you have any issues with the builds using versions that you are at in any given release.

**Comment by chrissimpkins** [comment]
@spstarr @anthrotype @davelab6 

> deep dive for typeface compilation / post build modification / font editors support on your distros

perhaps a `typeface-dev` package?  👍

**Comment by spstarr** [comment]
> perhaps a typeface-dev package?  :+1:

We try to package everything self-contained vs using pip install [module], this way we can ensure the files are always available (as the RPM contains them) and that we don't mess up system with non-packaged files.

**Comment by chrissimpkins** [comment]
@spstarr 

> We try to package everything self-contained vs using pip install [module]

Sorry, was referring to distro specific packages called `typeface-dev` that would install a set of validated open source typeface development tools at versions that would be self contained and distro specific.

```
$ sudo apt-get install typeface-dev
$ rpm -i typeface-dev
``` 

"One package to rule them all..." and the contained tools go through the same vetting process that all of your other packages do.  Short of this, it seems that for every new typeface that you want to support you potentially need to reinvent the wheel according to the developer rather than the developer attempting to adapt to your set of supported versions.  In cases where they cannot, individualized builds can be crafted with self contained dependencies but they will require the review process for new tools and all dependencies according to your build guidelines that is happening here.  May save significant effort and time for all involved.

**Comment by spstarr** [comment]
Fedora does have groups, there is no Font Development group, if someone requested that, all relevant packages for Font development could be installed for developers. There are also some packages that are metapackages but for pulling in the same package's own sub-packages.

**Comment by chrissimpkins** [comment]
As of https://github.com/source-foundry/Hack/commit/9284d227b0fbb48793e119914587df3d1af38c31 we have scripts to complete the build process from:

- UFO to ttf (desktop)
- UFO to woff (web) - includes full set and subset glyphs
- UFO to woff2 - includes full set and subset glyphs

Please let me know if you need any information about dependencies across these builds.

The web font builds are located in the `build-woff.sh`, `build-woff2.sh`, and `build-subsets.sh` files.

**Comment by chrissimpkins** [comment]
Builds will take place with the following set of commands (as of now):

Build desktop and web fonts (including subsets) with automated installation of all dependencies:

```
$ make build-with-dependencies
```

After build dependency installs (or if already installed), all future builds of all fonts are with:

```
$ make
```

Desktop ttf files:

```
$ make ttf
```

woff files (full set)

```
$ make woff
```

woff2 files (full set)

```
$ make woff2
```

woff and woff2 subsets

```
$ make subsets
```

All of the approaches that have been implemented to date are passing build tests on Travis across both Linux and OS X platforms.  :+1:

**Comment by chrissimpkins** [comment]
We are going to add a Python git library dependency to the build process in order to stamp our build version strings and web font CSS files with short git commit SHA1 strings.  This will be an optional step in the builds outside of this upstream repository and is intended for testing purposes on our end.  Beyond identification of the repository state at the time of the build, it will also indicate whether the fonts were built from a commit in this repo or from modified source outside of the repo.

Discussion about the approach is taking place in #293   

For packagers who are building from source, there is no need to place this new library (yet to be determined, will likely be GitPython)  in the review process for build dependencies unless you wish to follow the same approach for your own testing purposes with source modifications that occur in your downstream builds.

**Comment by chrissimpkins** [comment]
Estimate ~ 2 weeks to v3.0 release.

**Comment by spstarr** [comment]
An Update on dependencies for python-fontmake:

python-setuptools_scm_git_archive has been approved in Rawhide now
python-booleanoperations is in review, should be approved soon hopefully.
python-pyclipper is taking time, packager is working with upstream on de-bundling this

Once these clear, python-fontmake will be ready for review and I'll give it review then we can get onto getting Hack 3.0 in.

**Comment by chrissimpkins** [comment]
@spstarr excellent!  good news!  Thanks for the update

**Comment by chrissimpkins** [comment]
For those who re-distribute with our compiled fonts via Github releases, we will be adding tar.gz archives to the releases.  The following archives will be supported:

- zip
- tar.gz
- tar.xz

The current archive file path format will be:

```
Hack-v3.000-ttf.zip
Hack-v3.000-ttf.tar.gz
Hack-v3.000-ttf.tar.xz
Hack-v3.000-webfonts.zip
Hack-v3.000-webfonts.tar.gz
Hack-v3.000-webfonts.tar.xz
```

Issue report #325 if you'd like to discuss.  Note slight file path change from previous format:

```
Hack-v3_000-ttf.zip
```

**Comment by chrissimpkins** [comment]
Hack v3.000 PR is now available for review.  Goal for release ~ 2 weeks.

https://github.com/source-foundry/Hack/pull/335

**Comment by chrissimpkins** [comment]
v3.000 release is now merged to master and live.  Thanks all for your feedback and help with this major release!

**Comment by spstarr** [comment]
We are still blocked in Fedora, I'll see if I can intervene in the package review where possible.

**Comment by chrissimpkins** [comment]
@spstarr thanks! let us know if there is anything that we can do on our end.  This an issue with build dependencies?

**Comment by spstarr** [comment]
Yes, python-booleanoperations, python-pyclipper dependencies for python-fontmake. The problem right now is python-pyclipper bundles systems libraries, upstream is looking at modifying their install to support the system libraries (see here: https://github.com/greginvm/pyclipper/issues/10). If that is the solution, then the packaging for this can just set the environment variable accordingly to use the system lib.

For python-booleanoperations, it is blocked on python-pyclipper...

I've responded to the bugzilla tickets to see if they need help pushing the reviews. Though for python-pyclipper, I'm hoping that will be resolved soon. Otherwise, Fedora could provide distribution specific patches to use the system libs to workaround this in meantime.

**Comment by chrissimpkins** [comment]
@spstarr thanks for all of your help! sorry it is such a hassle.

**Comment by spstarr** [comment]
Further below:

"Yes, it is already done. The problem is the the latest polyclipping ABI is
different from the one used in this python package. We should also patch it to
use the latest ABI."

https://bugzilla.redhat.com/show_bug.cgi?id=1440992

**Comment by davelab6** [comment]
> Fedora does have groups, there is no Font Development group,

There is the https://lists.fedoraproject.org/admin/lists/fonts.lists.fedoraproject.org/ tho

**Comment by chrissimpkins** [comment]
@spstarr do you want this issue report open until you complete the release process on Fedora?  We pulled this into a discussion about the new scripted build approach but this was originally your thread for the Fedora redistribution package.  Happy to reopen if this is helpful in any way.

**Comment by spstarr** [comment]
@davelab6 yeah, the mailing list, I meant there's no comps.xml group to automatically to install dependencies for fonts development in general.

**Comment by spstarr** [comment]
@chrissimpkins Let's leave it open please. Maybe we can rename this issue?

**Comment by chrissimpkins** [comment]
@davelab6 Dave, has there been any coordinated effort on the part of OS tool development teams to package up a set of "best practices" compilation/post-compile modification tools for the Linux distros?  If not, do you think that this is something worthy of further discussion so that someone/a group of someones can help to establish a consistent set of commonly used tools across all interested Linux distros in order to avoid this issue for other typeface projects out there?  It seems that this is currently driven by interested distro users who donate their time to make project specific packages happen.  If we think a bit more generally about an approach, we could probably establish tooling that works for many projects rather than attempting to reinvent the wheel with each new typeface that wants to release on a Linux distro that has free software redistribution guidelines like Fedora/Debian/Ubuntu.

I am sure others have faced the issue, but I am guessing most are simply building with fontforge on Linux?  With the availability of all of the excellent OS typeface tooling out there, it might make sense for tool authors to become more involved in driving these projects through the Linux review processes.  The individuals who take the reins on that would have the opportunity to help define the build process standards on Linux distros given the time-intensive nature of these review processes for redistribution.  This definitely is a disincentive for all involved.

**Comment by chrissimpkins** [comment]
@spstarr opened again.  Feel free to rename however appropriate.

**Comment by chrissimpkins** [comment]
@spstarr any progress on the pyclipping library side with this issue?  Is this still where things are stuck?

**Comment by chrissimpkins** [comment]
@spstarr It doesn't appear that there has been any progress in the linked Fedora thread since October.  Is someone still working on this and if not can we close this issue report?

**Comment by bkmgit** [comment]
Building on Fedora can be partially done using
```
dnf -y install ttfautohint fonttools
pip install fontmake
git clone https://github.com/source-foundry/Hack
cd Hack
```
then as indicated at https://github.com/ryanoasis/nerd-fonts/issues/70#issuecomment-373959508 change the contents of the files in `postbuild_processing/tt-hinting/` so that the name of the glyph rather than the unicode number is used, for example for `Hack-Bold-TA.txt`
```
# U+0021 exclam glyph ID 580
exclam touch 22,23,24,25    y -0.5    @14

# U+0025 percent glyph ID 762
percent touch 0,1,16                     y 0.75   @10,11
percent touch 23,24,25                   y 0.25   @10,11
percent touch 17,18,32,46,47,48          y 0.5    @10,11
percent touch 57,58,71                   y -0.25  @10,11
percent touch 33,34,35,36                y 0.5    @10,11
percent touch 63,64,65                   y 0.75   @10,11

percent touch 23,24,25,63,64,65          y 0.5    @14
percent touch 17,18,32,57,58,71          y -0.5   @14

# U+002B plus glyph ID 765
plus  touch 0,1,2,3,6,7,8,9           y 0.5    @10,11

# U+0038 eight glyph ID 556
eight  touch 41,42,43                  y 0.25   @12,13,14
eight  touch 34,35,48                  y -0.25  @12,13,14
```
then build using
```
./build-ttf.sh --install-dependencies
```

**Comment by bkmgit** [comment]
When using the dev branch, the following works
```
dnf -y install ttfautohint fonttools pipenv
pip install fontmake
git clone https://github.com/source-foundry/Hack
cd Hack
git checkout dev
./build-ttf.sh --system
```
